CONTRIBUTING.md

```diff
@@ -49,15 +49,88 @@ If your pull request contains any of the below, it's less likely to be merged:
 **During pull request review**
 A core contributor will review your pull request and provide feedback. To ensure that there is not a large backlog of inactive PRs, the pull request will be marked as stale after two weeks of no activity. After another four days, it will be closed.
 
-### Architecture
-The architecture of project is enforced with validation diagrams and will be run automatically at build. See the `eng/DependencyValidation` project to update or adjust this diagram. The following is currently enforced:
+## Architecture
+
+To make sure that documentation is fresh, most explanations of Upgrade Assistant's architecture reside in in-code comments which version with the code. Below are the key components to be familiar with in Upgrade Assistant with links to their source files where in-code comments will provide more details on architectural decisions.
+
+### Startup and hosting
+
+- [`Program`](src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/Program.cs) is Upgrade Assistant's entry point. From there, command line arguments are parsed and user commands are executed.
+- [`ConsoleUpgradeCommand`](src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/Commands/Upgrade/ConsoleUpgradeCommand.cs) is the command that runs the upgrade workflow. It creates a `Microsoft.Extensions.Hosting` [generic host](https://docs.microsoft.com/dotnet/core/extensions/generic-host) to register and execute upgrade services. All services are resolved from the [dependency injection container](https://docs.microsoft.com/dotnet/core/extensions/dependency-injection) setup by the command, which makes it easy for extensions to register services in addition to those registered by default.
+- [`ConsoleUpgrade`](src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/Commands/Upgrade/ConsoleUpgrade.cs) is the `IAppCommand` that actually runs the REPL loop to upgrade the user's project.
+- [`IUpgradeStartup`](src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/IUpgradeStartup.cs) implementations are executed once as Upgrade Assistant starts up to run pre-commands or otherwise configure settings that will be needed by other components as the tool executes. Current built-in startup implementations include:
+  - [`MSBuildRegistrationStartup`](src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildRegistrationStartup.cs) which registers the MSBuild instance on the user's computer that will be used while running Upgrade Assistant.
```

Comment:
Startup steps may be grouped together since it's anything implementing IUpgradeStartup

Reply:
Yea, I have the IUpgradeStartup implementations listed here as sub-bullets of IUpgradeStartup. Is there more grouping/something different you think would be better here?